numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -1884,8 +1884,36 @@ gentype_reduce(PyObject *self, PyObject *NPY_UNUSED(args))
     PyTuple_SET_ITEM(ret, 0, obj);
     obj = PyObject_GetAttrString((PyObject *)self, "dtype");
     if (PyArray_IsScalar(self, Object)) {
-        mod = ((PyObjectScalarObject *)self)->obval;
-        PyTuple_SET_ITEM(ret, 1, Py_BuildValue("NO", obj, mod));
+        PyObject *val = ((PyObjectScalarObject *)self)->obval;
+        PyObject *tup = Py_BuildValue("NO", obj, val);
+        if (tup == NULL) {
+            return NULL;
+        }
+        PyTuple_SET_ITEM(ret, 1, tup);
+    }
+    else if (obj && PyDataType_FLAGCHK((PyArray_Descr *)obj, NPY_LIST_PICKLE)) {
+        /* a structured dtype with an object in a field */
+        PyObject *arr = PyArray_FromScalar(self, NULL);
+        if (arr == NULL) {
+            return NULL;
+        }
+        PyArrayIterObject *it = (PyArrayIterObject *)PyArray_IterNew(arr);
+        if (it == NULL) {
+            Py_DECREF(arr);
+            return NULL;
+        }
+        /* arr.item() */
+        PyObject *val = PyArray_GETITEM((PyArrayObject *)arr, it->dataptr);
```

Comment:
Har, fun dance, and all just because PyArray_GETITEM needs a (fake) array object. We need to fix it, but for now its good.

Reply:
Can you delete the iteterator? I do not understand why `PyArray_DATA` will not be just as good here. It should be a 1D array, no?